### PR TITLE
Release branch 3.2.0

### LINKED ISSUE #1727: `used-before-assignment` false negative after conditionals
Running pylint on this simple example does not detect that x can be used before assignment.  This is a very common programming mistake and should be caught.
```python
def a_function(d):
    if d:
        x=1
    return x
```

### LINKED ISSUE #2832: Warn about @contextlib.contextmanager without try/finally in generator functions
### Is your feature request related to a problem? 

If you decorate a generator function with `@contextlib.contextmanager` and use the context manager `cm` inside a generator function, the context manager's `__exit__` will not be always be called because the generator raises `GeneratorExit`, which ends the generator that defines `cm`. 

Notice that `cm exit` never appears, as `cm`'s `__exit__` is never called.

```python
from __future__ import generator_stop

import contextlib
import weakref
import sys


@contextlib.contextmanager
def cm():
    print("cm enter")
    a = yield
    print('cm exit')

def genfunc():
    with cm():
        print("stepping")
        yield

def main():
    gen = genfunc()
    ref = weakref.ref(gen, print)
    next(gen)


if __name__ == "__main__":
    main()

```
Output:

```
cm enter
stepping
<weakref at 0x7f4739e38e08; dead>
```

### Describe the solution you'd like

Generator functions decorated with `@contextlib.contextmanager` should not be used in generators unless the context manager has `finally:`

```python
@contextlib.contextmanager
def cm():
    try:
        yield
    finally:
        # cleanup code here
        ...
```

or specifically handles `GeneratorExit`:

```python
@contextlib.contextmanager
def cm():
    try:
        yield
    except GeneratorExit:
        # cleanup code here
        ...
```

### Additional context
Add any other context about the feature request here.

### LINKED ISSUE #3501: conditional import against six.PY2 emits error
### Steps to reproduce
1. Given following code
```
import six
if six.PY3:
    from urllib.parse import urlparse
else:
    from urlparse import urlparse
```
2. run pylint with python2

### Current behavior

```
[E0611(no-name-in-module), ] No name 'parse' in module 'urllib'
```

### Expected behavior
no error reported

### pylint --version output
```
Using config file .pylintrc
pylint 1.9.5,
astroid 1.6.5
Python 2.7.16 (default, Jul 12 2019, 01:42:49)
[GCC 6.3.0 20170516]
```

ref: https://github.com/PyCQA/pylint/issues/769

### LINKED ISSUE #4696: methods that return a sequence yield unpacking-non-sequence
If you unpack a sequence return value of a method, pylint falsely reports unpacking-non-sequence. It seems to work correctly with functions.

The issue is introduced with `pylint==2.9.0` and happens with all later versions, including the current `pylint==3.0.0a4`. It does **not** happen with `pylint==2.8.3`, which is why I think this is a different issue from the similar looking #4137 (which mentions 2.7.0).

### Steps to reproduce

The below example shows a minimal showcase of the false positive for methods and also shows that the error does not happen with functions.

```py
# pylint: disable=missing-module-docstring,missing-function-docstring,unused-argument,missing-class-docstring
import typing


@typing.overload
def func_overload(arg: str) -> typing.Tuple[str, str]:
    ...


@typing.overload
def func_overload(arg: bytes) -> typing.Tuple[bytes, bytes]:
    ...


def func_overload(arg: typing.Union[str, bytes]) -> typing.Tuple[str, str]:
    return "x", "y"


class Test:
    @typing.overload
    def meth_overload(self, arg: str) -> typing.Tuple[str, str]:
        ...

    @typing.overload
    def meth_overload(self, arg: bytes) -> typing.Tuple[bytes, bytes]:
        ...

    def meth_overload(self, arg: typing.Union[str, bytes]) -> typing.Tuple[str, str]:
        return "x", "y"


X, Y = func_overload("foo")  # OK
A, B = Test().meth_overload("foo")  # "E0633: Attempting to unpack a non-sequence (unpacking-non-sequence)"
```

When you invoke pylint on this file, the second line will show the false positive (and not the first one):

```
$ pylint test.py
************* Module test
test.py:33:0: E0633: Attempting to unpack a non-sequence (unpacking-non-sequence)

------------------------------------------------------------------
Your code has been rated at 6.88/10 (previous run: 6.88/10, +0.00)
```

### LINKED ISSUE #9442: Azure import try/except over doubles pylint memory usage
### Bug description

We run pylint on a reasonably large codebase, ~900 .py files, ~350k lines.
Recently we've needed to support two similar versions of the Azure SDK and as such made the change below in a single file.

from
```python
from azure.mgmt.network.v2022_01_01 import NetworkManagementClient
from azure.mgmt.network.v2022_01_01.models import Resource as NetworkResource
```
to
```python
try:
    from azure.mgmt.network.v2022_01_01 import NetworkManagementClient
    from azure.mgmt.network.v2022_01_01.models import Resource as NetworkResource
except ImportError:
    from azure.mgmt.network import NetworkManagementClient
    from azure.mgmt.network.v2022_07_01.models import Resource as NetworkResource
```

This change has increased pylint's run time but more importantly, over doubled resident memory usage! I've tried various combinations of disables and module ignores but cannot get anywhere close to previous figures without reverting the code change. Including attempting to ignore the azure module altogether with `--ignored-modules=azure`

Numbers below are taken from `/usr/bin/time -v python3 -m pylint package1 package2 -f colorized -r n -j 1`

Old code:
```
Elapsed (wall clock) time (h:mm:ss or m:ss): 5:13.65
Maximum resident set size (kbytes): 1589112
```

New code:
```
Elapsed (wall clock) time (h:mm:ss or m:ss): 6:30.88
Maximum resident set size (kbytes): 3993300
```

New code with `--ignored-modules=azure`:
```
Elapsed (wall clock) time (h:mm:ss or m:ss): 6:42.64
Maximum resident set size (kbytes): 3994000
```

To make this more generic, I'm running without a pylintrc file using the latest version from PyPI. This has the side effect of producing masses of warning/error output as usually we have a fair number of disables, however the issue described is still present in this state.

### Configuration

_No response_

### Command used

```shell
/usr/bin/time -v python3 -m pylint package1 package2 -f colorized -r n -j 1
and
/usr/bin/time -v python3 -m pylint --ignored-modules=azure package1 package2 -f colorized -r n -j 1
```

### Pylint output

_n/a_

### Expected behavior

Original run times and memory usage

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.3
Python 3.9.18 (main, Aug 25 2023, 13:20:14) 
[GCC 11.4.0]
```

### OS / Environment

Ubuntu 22.04

### Additional dependencies

```python
# non-Azure modules removed
azure-common==1.1.28
azure-core==1.26.4
azure-identity==1.12.0
azure-keyvault-certificates==4.7.0
azure-keyvault-secrets==4.7.0
azure-mgmt-automation==1.0.0
azure-mgmt-compute==29.1.0
azure-mgmt-core==1.4.0
azure-mgmt-keyvault==10.2.1
azure-mgmt-monitor==6.0.0
azure-mgmt-network==21.0.1
azure-mgmt-resource==23.0.0
azure-mgmt-storage==21.0.0
azure-storage-blob==12.16.0
msrestazure==0.6.4
```

### LINKED ISSUE #9443: GitHub actions output format
### Current problem

GitHub actions/scripts can output [a special log syntax](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-a-warning-message) to create warning/error messages directly in the GitHub UI. This is awesome, since you will see the warnings/errors in the summary of your last run and directly in the `Files changed` tab in pull requests (PRs). As an example, consider this image:

RuboCop (the linter/formatter for Ruby code) has an [**"GitHub Actions Formatter"**](https://docs.rubocop.org/rubocop/formatters.html#github-actions-formatter) directly included:

> Useful for GitHub Actions. Formats offenses as [workflow commands](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#setting-a-warning-message) to create annotations in GitHub UI.

You can use it like this:

```
rubocop --format github
```

and everything is ready to go.

### Desired solution

I've seen that Pylint allows to specify the [`--output-format`](https://pylint.pycqa.org/en/latest/user_guide/usage/output.html#output-options) command line flag and that a [new JSON2 reporter](https://github.com/pylint-dev/pylint/pull/8929) was added recently. I wonder if there is also a GitHub reporter available. I haven't found one yet after some digging around in search engines ;)

### Additional context

The feature could then be used as follows:

```
pylint --output-format=github
```

### LINKED ISSUE #9449: Nonsensical returns are not reported as `useless-return` if they are indented
### Bug description

```python
# pylint: disable=missing-module-docstring,missing-function-docstring


def function1(parameter):
    parameter.do()
    return  # this is reported as `useless-return` as expected


def function2(parameter):
    if parameter:
        return
    return  # this is not reported as `useless-return`


def function3(parameter):
    if parameter:
        pass
    else:
        return  # this is not reported as `useless-return`


def function4(parameter):
    if parameter:
        return
    else:  # this is reported as `no-else-return` as expected
        return  # this is not reported as `useless-return`


def function5(parameter):
    try:
        parameter.do()
    except RuntimeError:
        parameter.other()
        return  # this is not reported as `useless-return`


def function6(parameter):
    try:
        parameter.do()
    except RuntimeError:
        return  # maybe even this should be reported - `pass` can be used
```

### Configuration

_No response_

### Command used

```shell
pylint a.py
```

### Pylint output

```python
************* Module a
4:0: R1711: Useless return at end of function or method (useless-return)
23:4: R1705: Unnecessary "else" after "return", remove the "else" and de-indent the code inside it (no-else-return)
```

### Expected behavior

Maybe more `useless-return` messages should be reported, see the comments in the code snippet.

### Pylint version

```shell
pylint 2.17.7
astroid 2.15.8
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### LINKED ISSUE #9584: `property-with-parameters` only cares about non-`self` parameters if they're positional-or-keyword parameters
### Bug description

Pylint's `property-with-parameters` rule checks for `@property`-decorated functions that have non-`self` parameters, on the grounds that additional parameters in a property function can never be used. However, the check currently only seems to care about additional parameters if they are positional-or-keyword parameters. The check does not emit any violations for additional positional-only parameters, additional keyword-only parameters, additional variadic positional parameters, or additional variadic keyword parameters. This seems off to me: defining those additional parameters is just as strange/incorrect as defining additional positional-or-keyword parameters.

To illustrate, only the `a` property in the following example triggers an `R0206` violation, whereas I would expect `R0206` to be emitted on `a`, `b`, `c`, `d` and `e`:

```py
class Foo:
   @property
   def a(self, arg):
       return arg

   @property
   def b(self, arg, /):
      return arg

   @property
   def c(self, *, arg):
      return arg

   @property
   def d(self, *args):
      return args

   @property
   def e(self, **kwargs):
      return kwargs
```

This means that `R0206` is not actually emitted on the motivating example originally given for this check in https://github.com/pylint-dev/pylint/issues/3006#issue-468017491, since that example has an additional variadic keyword parameter (`**kwargs`).

### Command used

```shell
pylint foo.py --disable=W0311,C0114,C0104,C0115,C0116,R0903
```

### Pylint output

```
************* Module foo
foo.py:3:3: R0206: Cannot have defined parameters for properties (property-with-parameters)
```

### Expected output

```
************* Module foo
foo.py:3:3: R0206: Cannot have defined parameters for properties (property-with-parameters)
foo.py:7:3: R0206: Cannot have defined parameters for properties (property-with-parameters)
foo.py:11:3: R0206: Cannot have defined parameters for properties (property-with-parameters)
foo.py:15:3: R0206: Cannot have defined parameters for properties (property-with-parameters)
foo.py:19:3: R0206: Cannot have defined parameters for properties (property-with-parameters)
```

### Pylint version

```shell
pylint 3.1.0
astroid 3.1.0
Python 3.12.2 (main, Feb 15 2024, 19:30:27) [Clang 15.0.0 (clang-1500.1.0.2.5)]
```

### OS / Environment

MacOS

### Additional dependencies

None

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
What's new in Pylint 3.2.0?
---------------------------
Release date: 2024-05-14

New Features
------------

- Understand `six.PY2` and `six.PY3` for conditional imports.

  Closes #3501

- A new `github` reporter has been added. This reporter  returns the output of `pylint` in a format that
  Github can use to automatically annotate code. Use it with `pylint --output-format=github` on your Github Workflows.

  Closes #9443. 

New Checks
----------

- Add check ``possibly-used-before-assignment`` when relying on names after an ``if/else``
  switch when one branch failed to define the name, raise, or return.

  Closes #1727 

- Checks for generators that use contextmanagers that don't handle cleanup properly.
  Is meant to raise visibilty on the case that a generator is not fully exhausted and the contextmanager is not cleaned up properly.
  A contextmanager must yield a non-constant value and not handle cleanup for GeneratorExit.
  The using generator must attempt to use the yielded context value `with x() as y` and not just `with x()`.

  Closes #2832

False Negatives Fixed
---------------------

- If and Try nodes are now checked for useless return statements as well.

  Closes #9449. 

- Fix false negative for ``property-with-parameters`` in the case of parameters which are ``positional-only``, ``keyword-only``, ``variadic positional`` or ``variadic keyword``.

  Closes #9584 

False Positives Fixed
--------------------

- pylint now understands the ``@overload`` decorator return values better.

  Closes #4696
  Refs #9606

Performance Improvements
------------------------

- Ignored modules are now not checked at all, instead of being checked and then
  ignored. This should speed up the analysis of large codebases which have
  ignored modules.

  Closes #9442 

- ImportChecker's logic has been modified to avoid context files when possible. This makes it possible
  to cache module searches on astroid and reduce execution times.

  Refs #9310. 

- An internal check for ``trailing-comma-tuple`` being enabled for a file or not is now
  done once per file instead of once for each token.

  Refs #9608.

### CODE REVIEW

File: doc/whatsnew/3/3.2/index.rst [hunk 1]
```diff
@@ -12,5 +12,77 @@
...
+
+- Fix false negative for ``property-with-parameters`` in the case of parameters which are ``positional-only``, ``keyword-only``, ``variadic positional`` or ``variadic keyword``.
+
+  Closes #9584 (`#9584 <https://github.com/pylint-dev/pylint/issues/9584>`_)
+
+
+Internal Changes
+----------------
+
+- Update astroid version to 3.2.0.
```
Comment (jacobtylerwalls) [inline doc/whatsnew/3/3.2/index.rst:68]: I always wonder about shouting about the astroid changes. One major improvement people were waiting for was the fix for `@overload`. Do you want to mention it here someplace?

Comment (Pierre-Sassoulas) [reply doc/whatsnew/3/3.2/index.rst:68]: Yeah translating that into the implication for pylint would be better for sure.

**Comment by Pierre-Sassoulas** [comment]
@jacobtylerwalls did you see #9609 ?  How about we release the performance part (https://github.com/pylint-dev/pylint/pull/9609/commits/14af968ee22685d5c77189a1c9d355438d5aae62) without the bug fix part and with a warning about known issue ? This is going to make pylint faster and not more buggy than the last available version

File: doc/whatsnew/3/3.2/index.rst (hunk 1 above)
Comment (Pierre-Sassoulas) [reply doc/whatsnew/3/3.2/index.rst:68]: How about that (in false positives fixed):

```
pylint now understand the ``@overload`` decorator return values better.

Closes #4696
```

Comment (jacobtylerwalls) [reply doc/whatsnew/3/3.2/index.rst:68]: I dig it. I'd go with "understands" 👍

**Comment by jacobtylerwalls** [comment]
Sure, that sounds reasonable.

**Comment by Pierre-Sassoulas** [comment]
Opened https://github.com/pylint-dev/pylint/pull/9620, I'll upgrade the release note once it's merged.

**Comment by Pierre-Sassoulas** [comment]
Let's release then 😄 ! Do you want to do it @jacobtylerwalls ?

**Review by jacobtylerwalls — APPROVED** [review]
Looks good, I edited the markdown in the PR to use a bullet point for the astroid update. Not worth changing the .rst file.

**Pull request merged**